<a href="https://colab.research.google.com/github/Kavya2099/mlbookcamp/blob/main/Exercise_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

get data

In [80]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

--2022-09-24 17:09:05--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv.5’

housing.csv.5       100%[===================>]   1.36M  --.-KB/s    in 0.07s   

2022-09-24 17:09:06 (19.6 MB/s) - ‘housing.csv.5’ saved [1423529/1423529]



In [81]:
import pandas as pd
import numpy as np

df = pd.read_csv("housing.csv")
df.head()
df1 = df.copy()

In [82]:
df.isnull().sum()

df = df.fillna(0)

In [83]:
df['rooms_per_household']= df.total_rooms/df.households
df['bedrooms_per_room']= df.total_bedrooms/df.total_rooms
df['population_per_household']= df.population/df.households

What is the most frequent observation (mode) for the column ocean_proximity?

In [84]:
df.ocean_proximity.mode()

0    <1H OCEAN
dtype: object

Correlation

In [85]:
X= df.drop(['median_house_value'],axis=1)

num_cols= [col for col in X.columns if X[col].dtype in ['int64','float64']]
print(num_cols)

cat_cols = [col for col in X.columns if X[col].dtype == 'object']
print(cat_cols)

df[num_cols].corr()

['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'rooms_per_household', 'bedrooms_per_room', 'population_per_household']
['ocean_proximity']


longitude  latitude  housing_median_age  \
longitude                  1.000000 -0.924664           -0.108197   
latitude                  -0.924664  1.000000            0.011173   
housing_median_age        -0.108197  0.011173            1.000000   
total_rooms                0.044568 -0.036100           -0.361262   
total_bedrooms             0.068082 -0.065318           -0.317063   
population                 0.099773 -0.108785           -0.296244   
households                 0.055310 -0.071035           -0.302916   
median_income             -0.015176 -0.079809           -0.119034   
rooms_per_household       -0.027540  0.106389           -0.153277   
bedrooms_per_room          0.084836 -0.104112            0.125396   
population_per_household   0.002476  0.002366            0.013191   

                          total_rooms  total_bedrooms  population  households  \
longitude                    0.044568        0.068082    0.099773    0.055310   
latitude                    -0.036100       -0.065318   -0.108785   -0.071035   
housing_median_age          -0.361262       -0.317063   -0.296244   -0.302916   
total_rooms                  1.000000        0.920196    0.857126    0.918484   
total_bedrooms               0.920196        1.000000    0.866266    0.966507   
population                   0.857126        0.866266    1.000000    0.907222   
households                   0.918484        0.966507    0.907222    1.000000   
median_income                0.198050       -0.007295    0.004834    0.013033   
rooms_per_household          0.133798        0.002717   -0.072213   -0.080598   
bedrooms_per_room           -0.174583        0.122205    0.031397    0.059818   
population_per_household    -0.024581       -0.028019    0.069863   -0.027309   

                          median_income  rooms_per_household  \
longitude                     -0.015176            -0.027540   
latitude                      -0.079809             0.106389   
housing_median_age            -0.119034            -0.153277   
total_rooms                    0.198050             0.133798   
total_bedrooms                -0.007295             0.002717   
population                     0.004834            -0.072213   
households                     0.013033            -0.080598   
median_income                  1.000000             0.326895   
rooms_per_household            0.326895             1.000000   
bedrooms_per_room             -0.573836            -0.387465   
population_per_household       0.018766            -0.004852   

                          bedrooms_per_room  population_per_household  
longitude                          0.084836                  0.002476  
latitude                          -0.104112                  0.002366  
housing_median_age                 0.125396                  0.013191  
total_rooms                       -0.174583                 -0.024581  
total_bedrooms                     0.122205                 -0.028019  
population                         0.031397                  0.069863  
households                         0.059818                 -0.027309  
median_income                     -0.573836                  0.018766  
rooms_per_household               -0.387465                 -0.004852  
bedrooms_per_room                  1.000000                  0.003047  
population_per_household           0.003047                  1.000000


0.966507 ---> total_bedrooms with households

creating above avg variable

In [86]:
avg = df.median_house_value.mean()
avg

206855.81690891474

In [87]:
df['median_house_value'] = (df.median_house_value > avg).astype('int')
#df.median_house_value = df['above_average'].copy()
#df.drop(['above_average'],axis=1)

df.head()
df.tail()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
20635       845.0       330.0         1.5603                   0   
20636       356.0       114.0         2.5568                   0   
20637      1007.0       433.0         1.7000                   0   
20638       741.0       349.0         1.8672                   0   
20639      1387.0       530.0         2.3886                   0   

      ocean_proximity  rooms_per_household  bedrooms_per_room  \
20635          INLAND             5.045455           0.224625   
20636          INLAND             6.114035           0.215208   
20637          INLAND             5.205543           0.215173   
20638          INLAND             5.329513           0.219892   
20639          INLAND             5.254717           0.221185   

       population_per_household  
20635                  2.560606  
20636                  3.122807  
20637                  2.325635  
20638                  2.123209  
20639                  2.616981

Splitting data

In [88]:
from sklearn.model_selection import train_test_split

X= df.drop(['median_house_value'],axis=1)
y= df.median_house_value

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)


#spliting numerical and categorical features

num_cols= [col for col in X.columns if X[col].dtype in ['int64','float64']]
print(num_cols)

cat_cols = [col for col in X.columns if X[col].dtype == 'object']
print(cat_cols)

['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'rooms_per_household', 'bedrooms_per_room', 'population_per_household']
['ocean_proximity']


In [89]:
df.dtypes

longitude                   float64
latitude                    float64
housing_median_age          float64
total_rooms                 float64
total_bedrooms              float64
population                  float64
households                  float64
median_income               float64
median_house_value            int64
ocean_proximity              object
rooms_per_household         float64
bedrooms_per_room           float64
population_per_household    float64
dtype: object

In [90]:
# MI score for categorical feature
mi= train_X
for col in mi.select_dtypes('object'):
  mi[col],_ = mi[col].factorize()

discrete_feature= mi.dtypes == int

from sklearn.feature_selection import mutual_info_classif


mi_scores = mutual_info_classif(mi,train_y,discrete_features=discrete_feature)
mi_scores= pd.Series(mi_scores, name='mi_scores',index=mi.columns)
mi_scores= mi_scores.sort_values(ascending=False)
print(round(mi_scores,2))

median_income               0.19
longitude                   0.15
latitude                    0.14
ocean_proximity             0.10
bedrooms_per_room           0.08
rooms_per_household         0.05
population_per_household    0.04
total_rooms                 0.02
housing_median_age          0.01
households                  0.01
population                  0.00
total_bedrooms              0.00
Name: mi_scores, dtype: float64


ocean_proximity    --->          0.10

Creating logistic regression model

In [91]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

X= df.drop(['median_house_value'],axis=1)
y= df.median_house_value


X= pd.get_dummies(X)

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)

model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
model.fit(train_X, train_y)

pred = model.predict(val_X)
print('ACCURACY: ',accuracy_score(val_y,pred))
print('ACCURACY: ',round(accuracy_score(val_y,pred),2))

a=0.8335755813953488

ACCURACY:  0.8335755813953488
ACCURACY:  0.83


Finding the least useful feature using the feature elimination technique.

In [92]:
#drop total_rooms

X= df.drop(['median_house_value','total_rooms'],axis=1)
y= df.median_house_value


X= pd.get_dummies(X)

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)

model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
model.fit(train_X, train_y)

pred = model.predict(val_X)
print('ACCURACY: ',accuracy_score(val_y,pred))

print("Difference: ",a-accuracy_score(val_y,pred))

ACCURACY:  0.8345445736434108
Difference:  -0.0009689922480620172


In [93]:
#drop total_bedrooms

X= df.drop(['median_house_value','total_bedrooms'],axis=1)
y= df.median_house_value


X= pd.get_dummies(X)

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)

model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
model.fit(train_X, train_y)

pred = model.predict(val_X)
print('ACCURACY: ',accuracy_score(val_y,pred))
print("Difference: ",a-accuracy_score(val_y,pred))

ACCURACY:  0.8335755813953488
Difference:  0.0


In [94]:
#drop population

X= df.drop(['median_house_value','population'],axis=1)
y= df.median_house_value


X= pd.get_dummies(X)

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)

model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
model.fit(train_X, train_y)

pred = model.predict(val_X)
print('ACCURACY: ',accuracy_score(val_y,pred))
print("Difference: ",a-accuracy_score(val_y,pred))

ACCURACY:  0.8204941860465116
Difference:  0.013081395348837233


In [95]:
#drop households

X= df.drop(['median_house_value','households'],axis=1)
y= df.median_house_value


X= pd.get_dummies(X)

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)

model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
model.fit(train_X, train_y)

pred = model.predict(val_X)
print('ACCURACY: ',accuracy_score(val_y,pred))
print("Difference: ",a-accuracy_score(val_y,pred))

ACCURACY:  0.8301841085271318
Difference:  0.0033914728682170603


Total bedrooms is the least useful feature

Ridge regression model

In [96]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
from math import sqrt

#df1 is the copy of df
df1 = df1.fillna(0)
X= df1.drop(['median_house_value','households'],axis=1)
y= df1.median_house_value


a= [0, 0.01, 0.1, 1, 10]
X= pd.get_dummies(X)

fulltrain_X, val_X, fulltrain_y, val_y =  train_test_split(X,y,random_state=42,test_size=0.20)
train_X, test_X, train_y, test_y =  train_test_split(fulltrain_X,fulltrain_y,random_state=42,test_size=0.25)

for i in a:
  model = Ridge(alpha=i, solver="sag", random_state=42)
  model.fit(train_X, train_y)

  pred = model.predict(val_X)

  print("MSE for ",i," :",round(sqrt(mean_squared_error(pred,val_y,squared=False)),3))

MSE for  0  : 327.893
MSE for  0.01  : 327.893
MSE for  0.1  : 327.893
MSE for  1  : 327.893
MSE for  10  : 327.893


Selecting smallest alpha 0.01 as we have multiple options